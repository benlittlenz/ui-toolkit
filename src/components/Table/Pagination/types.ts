export type PaginationProps = {
  paginationIconLastPage?: React.ReactNode;
  paginationIconFirstPage?: React.ReactNode;
  paginationIconNext?: React.ReactNode;
  paginationIconPrevious?: React.ReactNode;
};
