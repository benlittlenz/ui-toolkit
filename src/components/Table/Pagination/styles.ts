import styled from 'styled-components';

export const PaginationWrapper = styled.nav`
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
  align-items: center;
  box-sizing: border-box;
  padding-right: 8px;
  padding-left: 8px;
  width: 100%;
`;

export const PageList = styled.div`
  display: flex;
  align-items: center;
  border-radius: 4px;
  white-space: nowrap;
`;

export const Button = styled.button`
  position: relative;
  display: block;
  user-select: none;
  border: none;
`;

const Span = styled.span`
  flex-shrink: 1;
  user-select: none;
`;

export const Range = styled(Span)`
  margin: 0 24px;
`;
