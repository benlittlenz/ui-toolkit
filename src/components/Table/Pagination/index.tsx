import * as React from 'react';

import { defaultProps } from '../defaultProps';
import { PaginationSelect } from '../Select';
import { getNumberOfPages } from '../utils';

import { Button, Range, PageList, PaginationWrapper } from './styles';
import { PaginationProps } from './types';

export function Pagination({
  currentPage,
  rowsPerPage,
  rowCount,
  paginationIconFirstPage = defaultProps.paginationIconFirstPage,
  paginationIconLastPage = defaultProps.paginationIconLastPage,
  paginationIconNext = defaultProps.paginationIconNext,
  paginationIconPrevious = defaultProps.paginationIconPrevious,
  onChangeRowsPerPage,
}: PaginationProps): JSX.Element {
  const lastIndex = currentPage * rowsPerPage;
  const firstIndex = lastIndex - rowsPerPage + 1;
  const numberOfPages = getNumberOfPages(rowCount, rowsPerPage);
  const disabledLesser = currentPage === 1;
  const disabledGreater = currentPage === numberOfPages;
  console.log('currentPage', currentPage);
  console.log('numberOfPages', numberOfPages);
  const range =
    currentPage === numberOfPages
      ? `${currentPage} of ${numberOfPages}`
      : `${firstIndex} of ${rowCount}`;

  const handleRowChange = React.useCallback(
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      return onChangeRowsPerPage(Number(e.target.value), currentPage);
    },
    [currentPage, onChangeRowsPerPage]
  );

  const selectOptions = [5, 10, 25].map((num: number) => (
    <option key={num} value={num}>
      {num}
    </option>
  ));
  return (
    <PaginationWrapper>
      <PaginationSelect
        onChange={handleRowChange}
        defaultValue={rowsPerPage}
        // aria-label={options.rowsPerPageText}
      >
        {selectOptions}
      </PaginationSelect>
      <Range>{range}</Range>
      <PageList>
        <Button
          id="pagination-first-page"
          type="button"
          aria-label="First Page"
          aria-disabled={disabledLesser}
          disabled={disabledLesser}
        >
          {paginationIconFirstPage}
        </Button>
        <Button
          id="pagination-next-page"
          type="button"
          aria-label="Next Page"
          aria-disabled={disabledLesser}
          disabled={disabledLesser}
        >
          {paginationIconNext}
        </Button>
        <Button
          id="pagination-previous-page"
          type="button"
          aria-label="Previous Page"
          aria-disabled={disabledGreater}
          // onClick={handleNext}
          disabled={disabledGreater}
        >
          {paginationIconPrevious}
        </Button>

        <Button
          id="pagination-last-page"
          type="button"
          aria-label="Last Page"
          aria-disabled={disabledGreater}
          // onClick={handleNext}
          disabled={disabledGreater}
        >
          {paginationIconLastPage}
        </Button>
      </PageList>
    </PaginationWrapper>
  );
}
