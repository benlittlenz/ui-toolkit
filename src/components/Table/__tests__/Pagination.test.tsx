import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Pagination } from '../Pagination';

test('it should render with default props', () => {
  const { container } = render(
    <Pagination
      currentPage={1}
      rowsPerPage={10}
      rowCount={40}
      onPageChange={jest.fn()}
      onChangeRowsPerPage={jest.fn()}
    />
  );
  console.log(container.firstChild);
  expect(container.firstChild).toMatchSnapshot();
});

describe('when clicking the first page button', () => {
  test('it should call onChangePage with correct args when current page is greater than 1', () => {
    const onChangePageMock = jest.fn();

    const { container } = render(
      <Pagination
        currentPage={4}
        rowsPerPage={10}
        rowCount={40}
        onPageChange={onChangePageMock}
        onChangeRowsPerPage={jest.fn()}
      />
    );
    userEvent.click(container.querySelector('button#pagination-first-page') as HTMLButtonElement);

    expect(onChangePageMock).toHaveBeenCalled();
  });
  test('it should not call onChangePage when current page is 1', () => {
    const onChangePageMock = jest.fn();

    const { container } = render(
      <Pagination
        currentPage={1}
        rowsPerPage={10}
        rowCount={40}
        onPageChange={onChangePageMock}
        onChangeRowsPerPage={jest.fn()}
      />
    );
    userEvent.click(container.querySelector('button#pagination-first-page') as HTMLButtonElement);

    expect(onChangePageMock).toBeCalledTimes(0);
  });
});

describe('when clicking the last page', () => {
  test('it should call onChangePage with correct args when current page is less than last page', () => {
    const onChangePageMock = jest.fn();

    const { container } = render(
      <Pagination
        currentPage={1}
        rowsPerPage={10}
        rowCount={40}
        onPageChange={onChangePageMock}
        onChangeRowsPerPage={jest.fn()}
      />
    );

    userEvent.click(container.querySelector('button#pagination-last-page') as HTMLButtonElement);

    expect(onChangePageMock).toHaveBeenCalledTimes(1);
  });

  test('it should NOT call onChangePage with correct args when current page is the last page', () => {
    const onChangePageMock = jest.fn();

    const { container } = render(
      <Pagination
        currentPage={4}
        rowsPerPage={10}
        rowCount={40}
        onPageChange={onChangePageMock}
        onChangeRowsPerPage={jest.fn()}
      />
    );

    userEvent.click(container.querySelector('button#pagination-last-page') as HTMLButtonElement);

    expect(onChangePageMock).toHaveBeenCalledTimes(0);
  });

  describe('when clicking the next page button', () => {
    test('it should call onChangePage with correct args, when current page is not the last page', () => {
      const onChangePageMock = jest.fn();
      const { container } = render(
        <Pagination
          currentPage={1}
          rowsPerPage={10}
          rowCount={40}
          onPageChange={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      );
      userEvent.click(container.querySelector('button#pagination-next-page') as HTMLButtonElement);

      expect(onChangePageMock).toHaveBeenCalledTimes(1);
    });

    test('it should call onChangePage with correct args, when current page is the last page', () => {
      const onChangePageMock = jest.fn();
      const { container } = render(
        <Pagination
          currentPage={4}
          rowsPerPage={10}
          rowCount={40}
          onPageChange={onChangePageMock}
          onChangeRowsPerPage={jest.fn()}
        />
      );
      userEvent.click(container.querySelector('button#pagination-next-page') as HTMLButtonElement);

      expect(onChangePageMock).toHaveBeenCalledTimes(0);
    });
  });
});
