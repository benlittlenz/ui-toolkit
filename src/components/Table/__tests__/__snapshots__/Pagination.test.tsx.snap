// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render with default props 1`] = `
<nav
  class="sc-dkPtRN gfGbEA"
>
  <div
    class="sc-gsDKAQ jiKEFD"
  >
    <select
      class="sc-bdvvtL cGpzzO"
    >
      <option
        value="5"
      >
        5
      </option>
      <option
        selected=""
        value="10"
      >
        10
      </option>
      <option
        value="25"
      >
        25
      </option>
    </select>
    <svg
      aria-hidden="true"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="sc-jRQBWg sc-gKclnd fHFdDi jWkYZz"
  >
    1 of 40
  </span>
  <div
    class="sc-hKwDye mvyID"
  >
    <button
      aria-disabled="true"
      aria-label="First Page"
      class="sc-eCImPb fBSgP"
      disabled=""
      id="pagination-first-page"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        style="height: 12px; width: 12px;"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15.707 15.707a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 010 1.414zm-6 0a1 1 0 01-1.414 0l-5-5a1 1 0 010-1.414l5-5a1 1 0 011.414 1.414L5.414 10l4.293 4.293a1 1 0 010 1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      aria-disabled="true"
      aria-label="Previous Page"
      class="sc-eCImPb fBSgP"
      disabled=""
      id="pagination-previous-page"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        style="height: 12px; width: 12px;"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      aria-disabled="false"
      aria-label="Next Page"
      class="sc-eCImPb fBSgP"
      id="pagination-next-page"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        style="height: 12px; width: 12px;"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      aria-disabled="false"
      aria-label="Last Page"
      class="sc-eCImPb fBSgP"
      id="pagination-last-page"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        style="height: 12px; width: 12px;"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.293 15.707a1 1 0 010-1.414L14.586 10l-4.293-4.293a1 1 0 111.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M4.293 15.707a1 1 0 010-1.414L8.586 10 4.293 5.707a1 1 0 011.414-1.414l5 5a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</nav>
`;
